### PR TITLE
Decompose runner.Runner

### PR DESCRIPTION
**Pull request by dbentley**
runner.Runner was unwieldy and hard to implement/understand/use. Break it down.

This implements the API from https://github.com/scootdev/scoot/pull/167 (and fixes all references to new names)

Names that could be better:
  Statuses
  ServiceFacade

### CODE REVIEW

File: config/scootconfig/modules.go [hunk 1]
```diff
@@ -129,7 +129,7 @@ func (c *WorkersThriftConfig) Create(
 		di := dialer.NewSimpleDialer(tf, pf)
 		cl, _ := client.NewSimpleClient(di, string(node.Id()))
 		q := runners.NewPollingService(cl, cl, cl, pollingPeriod)
-		return workers.NewServiceWorker(q, timeout, defaultOverhead)
+		return workers.NewServiceWorker(q, taskTimeout, defaultOverhead)
```
Comment (jasonbs10) [inline config/scootconfig/modules.go]: Not necessarily the PR for it, but: what is the right place to clarify that a task has to reach a PREPARING/RUNNING state before taskTimeout is started? When we use timeout in smoketest, iirc, we expect a task to be canceled regardless of queued status. Should we add a way to specify when timeout starts ticking?

File: runner/runners/statuses.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+// NewRun creates a new RunID in state Preparing
+func (s *Statuses) NewRun() (runner.RunStatus, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	id := runner.RunID(strconv.FormatInt(s.nextRunID, 10))
+	s.nextRunID++
+
+	st := runner.RunStatus{
```
Comment (jasonbs10) [inline runner/runners/statuses.go]: Reminder, if PR-176 is accepted in its current state, we will need to initialize StdoutRef/StderrRef here.

File: runner/runners/facade.go [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package runners
+
+import (
+	"github.com/scootdev/scoot/runner"
+)
+
+// ServiceFacade makes a Service from component parts.
```
Comment (dgassaway) [inline runner/runners/facade.go]: Why not just RunnerService? Facade implies testing usage to me (as does file name)

File: runner/runners/facade.go [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package runners
```
Comment (dgassaway) [inline runner/runners/facade.go]: Can we put a brief comment header on all of the runner/runners/*.go implementation and other files

File: runner/README.md [hunk 5]
```diff
@@ -1,11 +1,57 @@
...
+## Statuses interfaces (runner/statuses.go) ##
+There are several interfaces having to do with reading/writing RunStatus'es.
+
+StatusWriter for writing Statuses.
+
+There are more reader interfaces:
+* StatusReader includes all the ones below
+* LegacyStatusReader offers legacy APIs
+* StatusQuerier offers Query, which allows waiting for some period of time
+* StatusQuerierNow offers QueryNow which is the same as Query, but does not allow waiting
```
Comment (dgassaway) [inline runner/README.md]: Update readme interface names (e.g. StatusQuerierNow vs StatusQueryNower)

File: runner/README.md [hunk 6]
```diff
@@ -1,11 +1,57 @@
...
+* tenancy (run one at a time? Or multiple? Fixed, or based on system utilization?)
+* queueing (should we accept work we can't start right now)
 
-There will be an Execer that calls os/exec to exec.
\ No newline at end of file
+When a Controller wants to start a Run of a Command, it will use an Invoker.
+
+For now we have one implementation, QueueController
+
+## Statuses interfaces (runner/statuses.go) ##
```
Comment (dgassaway) [inline runner/README.md]: Suggestions - StatusManager, StatusReaderWriter

File: runner/README.md [hunk 7]
```diff
@@ -1,11 +1,57 @@
...
+There are several interfaces having to do with reading/writing RunStatus'es.
+
+StatusWriter for writing Statuses.
+
+There are more reader interfaces:
+* StatusReader includes all the ones below
+* LegacyStatusReader offers legacy APIs
```
Comment (dgassaway) [inline runner/README.md]: Somewhat off topic, but I think we should be very aggressive in deleting all vestigial traces of legacy APIs quickly. Cleaning this out will never be easier than it is now.  I don't remember exactly what this is supporting - closed source binaries?

File: config/scootconfig/modules.go (hunk 1 above)
Comment (dbentley) [inline config/scootconfig/modules.go]: Ugh, yeah, we should figure that out. Added a TODO in runner.go

File: runner/runners/statuses.go (hunk 2 above)
Comment (dbentley) [inline runner/runners/statuses.go]: Commented in task_runner.go that we need to reenable this.

(As discussed with Jason off-line, turning this off for now to make this PR easier to land)

File: runner/runners/facade.go (hunk 4 above)
Comment (dbentley) [inline runner/runners/facade.go]: Added a bunch of comments at the top of files whose purpose is non-obvious; wdyt?

File: runner/runners/facade.go (hunk 3 above)
Comment (dbentley) [inline runner/runners/facade.go]: Renamed to Service.

File: runner/README.md (hunk 6 above)
Comment (dbentley) [inline runner/README.md]: Renamed Statuses to StatusManager

File: runner/README.md (hunk 7 above)
Comment (dbentley) [inline runner/README.md]: Our Thrift API is still in terms of all Statuses and Erase'ing. Added a comment to worker.thrift.

File: runner/README.md (hunk 5 above)
Comment (dbentley) [inline runner/README.md]: Done.

File: runner/runners/facade.go (hunk 4 above)
Comment (dgassaway) [inline runner/runners/facade.go]: Lgtm.  Last thing mv facade.go service.go?

Comment (dbentley) [inline runner/runners/facade.go]: Had already done, but had forgotten to push. Now pushed.

**Pull request merged**